### PR TITLE
Make scheduled executor service tasks robust against exceptions (#24575)

### LINKED ISSUE #24575: Hanging output rendering after exception during rendering
While trying to build a reproducer for #24370, I also hit the problem, that the output rendering stalls if there was an exception during rendering.
I got a `NegativeArraySizeException` during the rendering with Gradle 8.0.2. (Will open a separate issue about it)

The problem is, that the rendering loop in `ThrottlingOutputEventListener#scheduleUpdateNow` is a `ScheduledExecutorService` task that does not handle exceptions and should run every 100 ms.
But such tasks do not do any further repetitions as soon as an exception was not handled.
And thus due to the exception being unhandled, output stalls.

### PR DESCRIPTION
**Pull request by Vampire**
Fixes #24575

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- n/a Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- n/a Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [x] Verify design and implementation 
- [x] Verify test coverage and CI build status
- [x] Verify documentation

### CODE REVIEW

**Comment by jbartok** [comment]
Thank you for your interest in Gradle!

This PR needs a decision from the team responsible for that area. They have been informed, response time may vary.

File: subprojects/launcher/src/main/java/org/gradle/launcher/daemon/server/health/gc/GarbageCollectionCheck.java [hunk 1]
```diff
@@ -46,20 +49,26 @@ public GarbageCollectionCheck(Clock clock, GarbageCollectorMXBean garbageCollect
...
+                    nonHeapEvents.slideAndInsert(new GarbageCollectionEvent(clock.getCurrentTime(), memoryPoolMXBean.getUsage(), -1));
                 }
             }
-            if (memoryPoolMXBean.getType() == MemoryType.NON_HEAP && poolName.equals(nonHeapMemoryPool)) {
-                nonHeapEvents.slideAndInsert(new GarbageCollectionEvent(clock.getCurrentTime(), memoryPoolMXBean.getUsage(), -1));
-            }
+        } catch (Throwable t) {
+            // this class is used as task in a scheduled executor service, so it must not throw any throwable,
+            // otherwise the further invocations of this task get automatically and silently cancelled
+            LOGGER.debug("Exception while checking garbage collection", t);
```
Comment (donat) [inline subprojects/launcher/src/main/java/org/gradle/launcher/daemon/server/health/gc/GarbageCollectionCheck.java:71]: ❓ I think this should be logged as a warning or as an error.

File: subprojects/logging/src/main/java/org/gradle/internal/logging/console/ThrottlingOutputEventListener.java [hunk 2]
```diff
@@ -58,7 +62,13 @@ private void scheduleUpdateNow() {
         executor.scheduleAtFixedRate(new Runnable() {
             @Override
             public void run() {
-                onOutput(new UpdateNowEvent(clock.getCurrentTime()));
+                try {
+                    onOutput(new UpdateNowEvent(clock.getCurrentTime()));
+                } catch (Throwable t) {
+                    // this class is used as task in a scheduled executor service, so it must not throw any throwable,
+                    // otherwise the further invocations of this task get automatically and silently cancelled
+                    LOGGER.debug("Exception while displaying output", t);
```
Comment (donat) [inline subprojects/logging/src/main/java/org/gradle/internal/logging/console/ThrottlingOutputEventListener.java:70]: Same here.

File: subprojects/process-services/src/main/java/org/gradle/process/internal/health/memory/DefaultMemoryManager.java [hunk 3]
```diff
@@ -148,8 +148,10 @@ public void run() {
                 }
                 JvmMemoryStatus jvm = jvmMemoryInfo.getJvmSnapshot();
                 jvmBroadcast.onJvmMemoryStatus(jvm);
-            } catch (Exception ex) {
-                LOGGER.debug("Failed to collect memory status: {}", ex.getMessage(), ex);
+            } catch (Throwable t) {
+                // this class is used as task in a scheduled executor service, so it must not throw any throwable,
+                // otherwise the further invocations of this task get automatically and silently cancelled
+                LOGGER.debug("Failed to collect memory status: {}", t.getMessage(), t);
```
Comment (donat) [inline subprojects/process-services/src/main/java/org/gradle/process/internal/health/memory/DefaultMemoryManager.java:154]: Same here.

**Comment by lptr** [comment]
My concern with this PR is that we are now catching `Throwable`s all over the place. Shouldn't we be catching `Exception`s instead?

**Comment by ljacomet** [comment]
I believe that's motivated both in the issue this PR addresses and in the inline comments to explain the pattern.

The only alternative would be to introduce an oversight mechanism that would allow Gradle to restart these scheduled tasks if they failed.

File: subprojects/launcher/src/main/java/org/gradle/launcher/daemon/server/health/gc/GarbageCollectionCheck.java (hunk 1 above)
Comment (ljacomet) [reply subprojects/launcher/src/main/java/org/gradle/launcher/daemon/server/health/gc/GarbageCollectionCheck.java:71]: I am not sure, these are purely internal errors that a user cannot do anything about. For me they are useful messages only when debugging a specific problem.

**Comment by Vampire** [comment]
I'm usually also hesitant to catch `Throwable` instead of `Exception`.
But as @ljacomet already said, I tried to clarify it in the code comments.
For a task that is used as a task in a scheduled executor service, you really have to catch any throwable or the task's further executions are silently canceled.
While the JavaDoc says "If any execution of the task encounters an exception", it actually means "If any execution of the task encounters a throwable".
So unless you add some logic to reschedule the task if a non-`Exception` `Throwable` is thrown, which also means you need to catch the `Throwable`, it really is the correct way for these tasks to catch `Throwable` or the furture executions are cancelled.

Depending on the actual `Error` thrown it might not make much difference. If for example an `OutOfMemoryError` was thrown, it is likely that shortly after the error is also thrown at some other place the same error will happen and stop the whole daemon. But in case the error really only came there and the build continues otherwise, the scheduled task would simply be not run anymore.

Whether the caught exceptions are logged on `debug`, `warn`, or `error` I do not have strong feelings.
Please clarify that with each other @ljacomet and @donat and tell me how you want it.
My take on this was, that the user is not really interested in these exceptions as long as the task runs again and does the memory collection / garbage collector checking / output displaying. And in case the further tasks also run on error and the actions do not work, it can be investigated using `--debug` and then seeing the errors.
But as I said, I have no strong feelings about either way.

File: subprojects/launcher/src/main/java/org/gradle/launcher/daemon/server/health/gc/GarbageCollectionCheck.java (hunk 1 above)
Comment (donat) [reply subprojects/launcher/src/main/java/org/gradle/launcher/daemon/server/health/gc/GarbageCollectionCheck.java:71]: 👍

**Comment by donat** [comment]
@bot-gradle test RFM

**Comment by Vampire** [comment]
@donat do you think the build failures are something I need to look at, or something else?

**Comment by donat** [comment]
@bot-gradle test RFM

**Comment by donat** [comment]
@Vampire let's see if it happens again. I'll check the new build failures and let you know.

**Comment by donat** [comment]
@bot-gradle test and merge

**Comment by donat** [comment]
@Vampire thank you for your contribution (and your patience too)!

**Pull request merged**